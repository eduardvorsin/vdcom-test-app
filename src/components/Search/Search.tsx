import React, { useState } from 'react';
import SearchIcon from '@mui/icons-material/Search';
import {
  Box,
  FormHelperText,
  IconButton,
  Input,
  SxProps,
  Theme,
  styled,
} from '@mui/material';
import { visuallyHiddenStyles } from '../../App';
import { isEmpty } from '../../helpers/validation';

const SearchForm = styled('form')(() => ({
  width: '100%',
}));

const SearchWrapper = styled(Box)(() => ({
  position: 'relative',
}));

const SearchButton = styled(IconButton)(({ theme }) => ({
  position: 'absolute',
  zIndex: '10',
  top: '50%',
  left: '16px',
  transform: 'translateY(-50%)',
  padding: '0px',
  svg: {
    width: '20px',
    height: '20px',
    fill: '#000000',
    transition: 'fill 0.15s ease',
  },
  [theme.breakpoints.up('md')]: {
    svg: {
      width: '28px',
      height: '28px',
    },
  },
  '&:hover > svg, &:focus > svg, &:focus-visible > svg': {
    fill: '#666666',
    transition: 'fill 0.15s ease',
  },
}));

const SearchLabel = styled('label')(() => ({
  ...visuallyHiddenStyles,
}));

const SearchInput = styled(Input)(({ theme }) => ({
  '&': {
    width: '100%',
  },
  '.MuiInput-input': {
    fontFamily: 'Roboto Condensed',
    fontSize: '20px',
    borderRadius: '6px',
    backgroundColor: '#F2F2F2',
    lineHeight: '1.17',
    height: 'auto',
    borderWidth: '2px',
    borderStyle: 'solid',
    borderColor: '#F2F2F2',
    transition: 'border-color 0.15s ease',
    padding: '8px 8px 8px 39px',
    [theme.breakpoints.up('md')]: {
      padding: '12px 12px 12px 51px',
    },
    '&:focus': {
      borderWidth: '2px',
      borderStyle: 'solid',
      borderColor: '#797979',
      transition: 'border-color 0.15s ease',
    },
    '&.error:focus': {
      borderColor: '#d32f2f',
    },
    '&::placeholder': {
      opacity: '1',
      color: '#797979',
    },
    '&.error::placeholder': {
      color: '#d32f2f',
    },
  },
  '&::after, &::before': {
    display: 'none',
  },
}));

type SearchFormElements = HTMLFormControlsCollection & {
  search: HTMLInputElement,
}

type SearchProps = {
  id: string,
  onChange: React.ChangeEventHandler<HTMLInputElement>,
  value: string,
  sx?: SxProps<Theme>,
};

const Search: React.FC<SearchProps> = ({
  id,
  value,
  onChange,
  sx,
}) => {
  const [error, setError] = useState<string>('');
  const searchChangeHandler: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    if (isEmpty(e.currentTarget.value)) {
      setError('before sending, you need to fill in the search field');
    } else {
      setError('');
    }

    onChange(e);
  };

  const searchSubmitHandler: React.FocusEventHandler<HTMLFormElement> = (e) => {
    e.preventDefault();

    const searchFormElements = e.currentTarget.elements as SearchFormElements;

    if (isEmpty(searchFormElements.search.value)) {
      setError('before sending, you need to fill in the search field');
      searchFormElements.search.focus();
    } else {
      setError('');
      e.currentTarget.submit();
    }
  };

  return (
    <SearchForm
      sx={sx}
      onSubmit={searchSubmitHandler}
      noValidate
    >
      <SearchWrapper
        role='search'
      >
        <SearchButton
          type='submit'
          disableRipple
        >
          <SearchIcon />
        </SearchButton>
        <SearchLabel
          htmlFor={id}
        >
          Search about documents
        </SearchLabel>
        <SearchInput
          inputProps={{
            className: `${error === '' ? '' : 'error'}`,
          }}
          name='search'
          required
          error={error !== ''}
          id={id}
          placeholder='Search by Name...'
          value={value}
          onChange={searchChangeHandler}
          aria-describedby={`${id}-helper-text`}
        />
      </SearchWrapper>

      {error && (
        <FormHelperText
          error
          id={`${id}-helper-text`}
          sx={{
            textDecoration: 'line-through',
            fontFamily: 'Roboto Condensed',
            fontSize: '18px',
          }}
        >
          {error}
        </FormHelperText>
      )}
    </SearchForm>
  );
};

export default Search;
