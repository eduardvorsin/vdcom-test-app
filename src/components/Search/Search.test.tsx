
import { render, screen, waitFor } from '@testing-library/react';

import userEvent from '@testing-library/user-event';
import React from 'react';
import Search from './Search';

describe('Search Tests', () => {
  it('renders correctly', () => {
    const mockFn = jest.fn<void, [e: React.ChangeEvent<HTMLInputElement>]>();
    render(
      <Search
        id='search'
        onChange={mockFn}
        value=''
      />,
    );

    expect(screen.getByTestId<HTMLFormElement>('search-form')).toBeInTheDocument();
    expect(screen.getByRole<HTMLDivElement>('search')).toBeInTheDocument();
  });

  it('should call the mock function when clicking on "Search about documents"', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [e: React.ChangeEvent<HTMLInputElement>]>();
    render(
      <Search
        id='search'
        onChange={mockFn}
        value=''
      />,
    );

    user.type(screen.getByRole<HTMLInputElement>('textbox'), 'new');

    await waitFor(() => {
      expect(mockFn).toHaveBeenCalledTimes(3);
    });
  });

  it('should show an error if the input field is empty when clicking on the submit button', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [e: React.ChangeEvent<HTMLInputElement>]>();
    render(
      <Search
        id='search'
        onChange={mockFn}
        value=''
      />,
    );

    user.click(screen.getByRole<HTMLButtonElement>('button'));

    expect(await screen.findByText<HTMLParagraphElement>('before sending, you need to fill in the search field')).toBeInTheDocument();
  });

  it('is a basic snapshot', () => {
    const mockFn = jest.fn<void, [e: React.ChangeEvent<HTMLInputElement>]>();
    render(
      <Search
        id='search'
        onChange={mockFn}
        value=''
      />,
    );

    expect(screen.getByTestId<HTMLFormElement>('search-form')).toMatchSnapshot();
  });

  it('is a snapshot with an error', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [e: React.ChangeEvent<HTMLInputElement>]>();
    render(
      <Search
        id='search'
        onChange={mockFn}
        value=''
      />,
    );

    user.click(await screen.findByRole<HTMLButtonElement>('button'));

    await screen.findByText<HTMLParagraphElement>('before sending, you need to fill in the search field');

    expect(screen.getByTestId<HTMLFormElement>('search-form')).toMatchSnapshot();
  });
});
