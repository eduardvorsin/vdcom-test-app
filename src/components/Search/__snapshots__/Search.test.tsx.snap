// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Tests is a basic snapshot 1`] = `
<form
  class="css-8atqhb"
  data-testid="search-form"
  novalidate=""
>
  <div
    class="MuiBox-root css-19sk4h4"
    role="search"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-123v15u-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="submit"
    >
      search
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </button>
    <label
      class="css-inngth"
      for="search"
    >
      Search about documents
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-15tv2wk-MuiInputBase-root-MuiInput-root"
      required=""
    >
      <input
        aria-describedby="search-helper-text"
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        id="search"
        name="search"
        placeholder="Search by Name..."
        required=""
        type="text"
        value=""
      />
    </div>
  </div>
</form>
`;

exports[`Search Tests is a snapshot with an error 1`] = `
<form
  class="css-8atqhb"
  data-testid="search-form"
  novalidate=""
>
  <div
    class="MuiBox-root css-19sk4h4"
    role="search"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-123v15u-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="submit"
    >
      search
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </button>
    <label
      class="css-inngth"
      for="search"
    >
      Search about documents
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-error Mui-focused css-15tv2wk-MuiInputBase-root-MuiInput-root"
      required=""
    >
      <input
        aria-describedby="search-helper-text"
        aria-invalid="true"
        class="MuiInputBase-input MuiInput-input error css-1x51dt5-MuiInputBase-input-MuiInput-input"
        id="search"
        name="search"
        placeholder="Search by Name..."
        required=""
        type="text"
        value=""
      />
    </div>
  </div>
  <p
    class="MuiFormHelperText-root Mui-error css-1nhp958-MuiFormHelperText-root"
    id="search-helper-text"
  >
    before sending, you need to fill in the search field
  </p>
</form>
`;
